### PR TITLE
Optimize Netty Frame Decoding

### PR DESCRIPTION
**Pull request by original-brownbear**
* We should not create a new wrapper object if there's no bytes in the `ByteBuf` 
* We should not create a new wrapped `ByteBuf` if it can't contain a message anyway because it doesn't even have enough bytes for a header left
 * This one is particularly unfortunate right now because for every fully read `ByteBuf` we still ran the decode loop once again (because of the missing update to `continueDecode` that I  now added) so without this change every message (if it's the only buffered message) cost an additional object here

### CODE REVIEW

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoder.java [hunk 1]
```diff
@@ -35,7 +35,7 @@
     @Override
     protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
         try {
-            boolean continueDecode = true;
+            boolean continueDecode = in.readableBytes() >= HEADER_SIZE;
             while (continueDecode) {
                 int messageLength = TcpTransport.readMessageLength(Netty4Utils.toBytesReference(in));
```
Comment (original-brownbear) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoder.java]: At the cost of some duplication (between Netty and NIO impls.) we could just remove creating the wrapping `BytesReference` just for reading the length here as well. I'd look into that in a follow up.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoder.java [hunk 2]
```diff
@@ -46,9 +46,11 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
                     if (messageLengthWithHeader > in.readableBytes()) {
                         continueDecode = false;
                     } else {
-                        final ByteBuf message = in.retainedSlice(in.readerIndex() + HEADER_SIZE, messageLength);
+                        final int readerIndex = in.readerIndex();
+                        final ByteBuf message = in.retainedSlice(readerIndex + HEADER_SIZE, messageLength);
                         out.add(message);
-                        in.readerIndex(in.readerIndex() + messageLengthWithHeader);
+                        in.readerIndex(readerIndex + messageLengthWithHeader);
+                        continueDecode = in.readableBytes() >= HEADER_SIZE;
```
Comment (ywelsch) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoder.java]: simpler to just write:

```
while (in.readableBytes() >= HEADER_SIZE) {
    int messageLength = TcpTransport.readMessageLength(Netty4Utils.toBytesReference(in));
    if (messageLength == -1) {
        break;
    } else {
        int messageLengthWithHeader = messageLength + HEADER_SIZE;
        // If the message length is greater than the network bytes available, we have not read a complete frame.
        if (messageLengthWithHeader > in.readableBytes()) {
            break;
        } else {
            final ByteBuf message = in.retainedSlice(in.readerIndex() + HEADER_SIZE, messageLength);
            out.add(message);
            in.readerIndex(in.readerIndex() + messageLengthWithHeader);
        }
    }
}
```

**Review by ywelsch — CHANGES_REQUESTED** [review]

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoder.java (hunk 2 above)
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoder.java]: Right sorry about that, fixed in 9d1f6a6

**Comment by original-brownbear** [comment]
@ywelsch fixed :)

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
thanks @ywelsch !

**Pull request merged**